File: src/main/java/eladkay/quaritum/common/rituals/AltarOfTheFallingStarDIagram.java
package eladkay.quaritum.common.rituals;

import com.google.common.collect.Lists;
import eladkay.quaritum.api.rituals.EnumRitualDuration;
import eladkay.quaritum.api.rituals.EnumRitualType;
import eladkay.quaritum.api.rituals.IRitual;
import eladkay.quaritum.common.item.soulstones.Awakened;
import eladkay.quaritum.common.item.soulstones.Dormant;
import eladkay.quaritum.common.lib.LibNames;
import net.minecraft.entity.item.EntityItem;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemStack;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.math.AxisAlignedBB;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;

import java.util.ArrayList;
import java.util.List;

public class AltarOfTheFallingStarDiagram implements IRitual {
    @Override
    public String getUnlocalizedName() {
        return "rituals.quaritum.altarofthefallingstar";
    }

    @Override
    public EnumRitualType getRitualType() {
        return EnumRitualType.DIAGRAM;
    }

    @Override
    public EnumRitualDuration getRitualDuration() {
        return EnumRitualDuration.INSTANT;
    }

    @Override
    public boolean runOnce(World world, EntityPlayer player, BlockPos pos) {
        List<EntityItem> list = world.getEntitiesWithinAABB(EntityItem.class, new AxisAlignedBB(pos.add(8, 8, 8).add(-8, -8, -8)));
        int total = 0;
        EntityPlayer receiver = null;
        for (EntityItem entity : list) {
            if (entity.getEntityItem().getItem() instanceof Dormant && entity.getEntityItem().getTagCompound().getBoolean(LibNames.TAG_ATTUNED)) {
                receiver = new EntityPlayer(world, new com.mojang.authlib.GameProfile(entity.getEntityItem().getTagCompound().getUniqueId(LibNames.TAG_UUID), entity.getEntityItem().getTagCompound().getString(LibNames.TAG_OWNER))) {
                    @Override
                    public boolean isSpectator() {
                        return false;
                    }

                    @Override
                    public boolean isCreative() {
                        return false;
                    }
                };
                entity.setDead();
            } else if (entity.getEntityItem().getItem() instanceof Awakened) {
                total += entity.getEntityItem().getTagCompound().getInteger(LibNames.TAG_ANIMUS);
                entity.setDead();
            }
        }
        if (receiver != null && total > 0) {
            receiver.getEntityData().setInteger(LibNames.TAG_ANIMUS_ON_ENTITY, 50);
        }
        return true;
    }

    @Override
    public boolean runDurable(World world, EntityPlayer player, BlockPos pos) {
        return false;
    }

    @Override
    public boolean canRitualRun(World world, EntityPlayer player, BlockPos pos, TileEntity tile) {
        List<EntityItem> list = world.getEntitiesWithinAABB(EntityItem.class, new AxisAlignedBB(pos.add(8, 8, 8).add(-8, -8, -8)));
        boolean a = false;
        boolean b = false;
        for (EntityItem entity : list) {
            if (entity.getEntityItem().getItem() instanceof Dormant && entity.getEntityItem().getTagCompound().getBoolean(LibNames.TAG_ATTUNED))
                a = true;
            else if (entity.getEntityItem().getItem() instanceof Awakened) b = true;
        }
        return a && b;
    }

    @Override
    public ArrayList<ItemStack> getRequiredItems() {
        return Lists.newArrayList();
    }

    @Override
    public String getCanonicalName() {
        return "Altar of the Falling Star";
    }
}

File: src/main/java/eladkay/quaritum/common/rituals/diagrams/AltarOfTheFallingStarDiagram.kt
package eladkay.quaritum.common.rituals.diagrams

import amerifrance.guideapi.api.IPage
import amerifrance.guideapi.page.PageText
import com.google.common.collect.ImmutableList
import com.teamwizardry.librarianlib.client.util.TooltipHelper
import eladkay.quaritum.api.animus.AnimusHelper
import eladkay.quaritum.api.animus.INetworkProvider
import eladkay.quaritum.api.animus.ISoulstone
import eladkay.quaritum.api.lib.LibBook
import eladkay.quaritum.api.rituals.IDiagram
import eladkay.quaritum.api.rituals.PositionedBlock
import eladkay.quaritum.api.rituals.PositionedBlockChalk
import eladkay.quaritum.common.book.ModBook
import eladkay.quaritum.common.book.PageDiagram
import eladkay.quaritum.common.core.ChatHelper
import eladkay.quaritum.common.item.ModItems
import net.minecraft.item.EnumDyeColor
import net.minecraft.item.ItemStack
import net.minecraft.tileentity.TileEntity
import net.minecraft.util.math.BlockPos
import net.minecraft.util.text.TextComponentTranslation
import net.minecraft.world.World
import java.util.*

class AltarOfTheFallingStarDiagram : IDiagram {

    override fun constructBook() {
        pages.add(PageText(TooltipHelper.local(LibBook.ENTRY_CIRCLE_OF_THE_FINAL_MOMENT_PAGE1)))
        val list = ArrayList<PositionedBlock>()
        buildChalks(list)
        pages.add(PageDiagram(list, ImmutableList.of<ItemStack>()))
        ModBook.register(ModBook.pagesDiagrams, LibBook.ENTRY_ALTAR_OF_THE_FALLING_STAR, pages, ItemStack(ModItems.awakened))
    }

    override fun getUnlocalizedName(): String {
        return "rituals.quaritum.altarofthefallingstar"
    }

    override fun hasRequiredItems(world: World?, pos: BlockPos, tile: TileEntity): Boolean {
        val stacks = IDiagram.Helper.stacksAroundAltar(tile, 4.0)
        for (stack in stacks) {
            if (stack.item is ISoulstone) return true
        }
        return false
    }

    override fun run(world: World, pos: BlockPos, te: TileEntity) {
        var flag = false

        val entities = IDiagram.Helper.entitiesAroundAltar(te, 4.0)

        var player: UUID? = null
        for (item in entities) {
            val stack = item.entityItem
            if (stack.item !is INetworkProvider || !(stack.item as INetworkProvider).isReceiver(stack))
                continue
            player = (stack.item as INetworkProvider).getPlayer(stack)
            break
        }
        if (player == null) return

        for (item in entities) {
            val stack = item.entityItem
            if (stack.item !is ISoulstone) continue
            val ss = stack.item as ISoulstone
            AnimusHelper.Network.addAnimus(player, ss.getAnimusLevel(stack))
            AnimusHelper.Network.addRarity(player, ss.getRarityLevel(stack))
            item.setDead()
            flag = true
        }
        if (flag)
            for (playerEntity in world.playerEntities)
                if (playerEntity.uniqueID == player)
                    ChatHelper.sendNoSpam2(playerEntity, TextComponentTranslation("misc.quaritum.rushOfEnergy"))
    }


    override fun canRitualRun(world: World?, pos: BlockPos, tile: TileEntity): Boolean {
        val entities = IDiagram.Helper.entitiesAroundAltar(tile, 4.0)
        for (entity in entities) {
            val stack = entity.entityItem
            if (stack.item is INetworkProvider && (stack.item as INetworkProvider).isReceiver(stack))
                return true
        }
        return false
    }

    override fun buildChalks(chalks: MutableList<PositionedBlock>) {
        chalks.add(PositionedBlockChalk(EnumDyeColor.PINK, BlockPos(0, 0, -2)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.WHITE, BlockPos(-1, 0, -1)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.WHITE, BlockPos(0, 0, -1)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.WHITE, BlockPos(1, 0, -1)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.PINK, BlockPos(-2, 0, 0)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.WHITE, BlockPos(-1, 0, 0)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.WHITE, BlockPos(1, 0, 0)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.PINK, BlockPos(2, 0, 0)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.WHITE, BlockPos(-1, 0, 1)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.WHITE, BlockPos(0, 0, 1)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.WHITE, BlockPos(1, 0, 1)))
        chalks.add(PositionedBlockChalk(EnumDyeColor.PINK, BlockPos(0, 0, 2)))
    }

    companion object {
        var pages: MutableList<IPage> = ArrayList()
    }
}
